src/crypto/ecdsa.rs

```diff
@@ -0,0 +1,137 @@
+    }
+}
+
+/// Verify an ECDSA signature.
+pub fn verify(
+    curve: &ECCCurve,
+    p: &[u8],
+    _hash: HashAlgorithm,
+    hashed: &[u8],
+    sig: &[Mpi],
+) -> Result<()> {
+    match *curve {
+        ECCCurve::P256 => {
+            let flen = 32;
+            ensure_eq!(sig.len(), 2);
+
+            let r = sig[0].as_bytes();
+            let s = sig[1].as_bytes();
+
+            ensure!(r.len() <= flen, "invalid R (len)");
+            ensure!(s.len() <= flen, "invalid S (len)");
+            ensure_eq!(p.len(), 2 * flen + 1, "invalid P (len)");
+            ensure_eq!(p[0], 0x04, "invalid P (prefix)");
+
+            let pk = p256::ecdsa::VerifyingKey::from_sec1_bytes(p)?;
+            let mut sig_bytes = vec![0u8; 2 * flen];
+            //
+            // add padding if the values were encoded short
+            sig_bytes[(flen - r.len())..flen].copy_from_slice(r);
+            sig_bytes[flen + (flen - s.len())..].copy_from_slice(s);
+
+            let sig = p256::ecdsa::Signature::from_bytes(&sig_bytes)?;
+
+            pk.verify_prehash(hashed, &sig)?;
+
+            Ok(())
+        }
+        ECCCurve::P384 => {
+            let flen = 48;
+            ensure_eq!(sig.len(), 2);
+
+            let r = sig[0].as_bytes();
+            let s = sig[1].as_bytes();
+
+            ensure!(r.len() <= flen, "invalid R (len)");
+            ensure!(s.len() <= flen, "invalid S (len)");
+            ensure_eq!(p.len(), 2 * flen + 1, "invalid P (len)");
+            ensure_eq!(p[0], 0x04, "invalid P (prefix)");
+
+            let pk = p384::ecdsa::VerifyingKey::from_sec1_bytes(p)?;
+            let mut sig_bytes = vec![0u8; 2 * flen];
+            //
+            // add padding if the values were encoded short
+            sig_bytes[(flen - r.len())..flen].copy_from_slice(r);
+            sig_bytes[flen + (flen - s.len())..].copy_from_slice(s);
+
+            let sig = p384::ecdsa::Signature::from_bytes(&sig_bytes)?;
+
+            pk.verify_prehash(hashed, &sig)?;
+
+            Ok(())
+        }
+        _ => unsupported_err!("curve {:?} for ECDSA", curve.to_string()),
+    }
+}
+
+/// Sign using ECDSA
+pub fn sign(
+    curve: &ECCCurve,
+    secret_key: &ECDSASecretKey,
+    _hash: HashAlgorithm,
+    digest: &[u8],
+) -> Result<Vec<Vec<u8>>> {
+    let d = secret_key.x.to_bytes_be();
+
+    let bytes = match curve {
+        ECCCurve::P256 => {
+            let secret = p256::ecdsa::SigningKey::from_bytes(&d)?;
+            let signature = secret.sign_prehash(digest)?;
+            signature.to_vec()
+        }
+        ECCCurve::P384 => {
+            let secret = p384::ecdsa::SigningKey::from_bytes(&d)?;
+            let signature = secret.sign_prehash(digest)?;
+            signature.to_vec()
+        }
+        _ => unsupported_err!("curve {:?} for ECDSA", curve),
+    };
+
+    let len = bytes.len() / 2;
+    let r = bytes[..len].to_vec();
+    let s = bytes[len..].to_vec();
```

Comment:
instead of converting to a vec twice, as the signature length is fixed for each variant could do sth like 
```rust
let r = vec![0u8; len];
r.copy_from_slice(&signature[..len]);
```

Reply:
Not quite. `signature` isn't an array here, it is a thin wrapper around GenericArray,  but we don't have access to internals. I changed this to `signature.split_bytes()`, then each component is converted using `to_vec()`.